app/controllers/concerns/current_user.rb

```diff
@@ -58,19 +58,16 @@ def authorize_project_deployer!
 
   # tested via access checks in the actual controllers
   def authorize_resource!
-    unauthorized! unless can?(resource_action, controller_name.to_sym)
+    scope = current_stage || current_project if controller_name.to_sym == :locks
+    unauthorized! unless can?(resource_action, controller_name.to_sym, scope)
```

Comment:
how about we get rid of this hack and override `def authorize_resource!` in locks_controller ?
... the all the resource finding logic can go in there too ...